File: Solutions/110-Balanced Binary Tree/110-Balanced Binary Tree.cpp
//Third fastest
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return getHeight(root) != -1;
    }
    
    int getHeight(TreeNode* root) {
        if (!root)
            return 1;
        int left = getHeight(root->left);
        int right = getHeight(root->right);
        
        if (left != -1 && right != -1 && abs(left - right) <= 1) {
            return max(left, right) + 1;
        }
        
        return -1;
    }
    
};

//Second fastest
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
      bool isBalanced(TreeNode* root) {
        if(root == nullptr)
            return true;
        return height(root) != -1;
    }
    int height(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        int lHeight = height(root->left);
        int rHeight = height(root->right);
        if(lHeight == -1 || rHeight == -1 || abs(lHeight - rHeight) > 1){
            return -1;
        }
        return 1 + max(lHeight, rHeight);
    }
};

//Fastest
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return helper(root).first;
    }
    
    std::pair<bool, int> helper(TreeNode* node) {
        if (!node) {
            return std::pair<bool, int>(true, 0);
        }
        auto left = helper(node->left);
        auto right = helper(node->right);
        auto balance = left.first && right.first && (abs(left.second - right.second) <= 1);
        auto height = max(left.second, right.second) + 1;
        return std::pair<bool, int>(balance, height);
    }
};

File: Solutions/110-Balanced Binary Tree/110-Balanced Binary Tree.py
#Third fastest
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        flag = True
        
        def f(node):
            nonlocal flag
            if not node:
                return 0
            l = f(node.left)
            r = f(node.right)
            if abs(l - r) > 1:
                flag = False
            return max(l, r) + 1
        
        f(root)
        return flag

#Second fastest
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        
        def helper(node):
            if not node:
                return 0,True
            
            l_height,l_corr = helper(node.left)
            if not l_corr:
                return -1,False
            
            r_height,r_corr = helper(node.right)
            if not r_corr:
                return -1,False
            
            return (1 + max(l_height,r_height), abs(l_height - r_height) <= 1)
        
        return helper(root)[1]

#Fastest
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

    
class Solution(object):
    def helper(self, root):
        if root == None:
            return 0, True
        hl, lb = self.helper(root.left)
        hr, rb = self.helper(root.right)
        if lb and rb and abs(hl-hr) <= 1:
            return max(hl,hr) + 1, True
        else:
            return -1, False

    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        h, is_b = self.helper(root)
        return is_b